contents/blog/clickhouse-materialized-columns.md

```diff
@@ -0,0 +1,126 @@
+---
+date: 2021-09-09
+title: Speeding up ClickHouse with Materialized Columns
+rootPage: /blog
+sidebar: Blog
+showTitle: true
+hideAnchor: true
+categories: engineering
+author: karl-aksel-puulmann
+featuredImage: ../images/blog/automating-software-company-github-actions.png
+featuredImageType: full
+---
+
+One quite obscure feature of ClickHouse is Materialized Columns, which allows dynamically creating columns using an expression. In this post, I’ll walk through an example query optimization in which materialized columns work well.
+
+Consider the following schema:
+
+```sql
+CREATE TABLE events (
+    uuid UUID,
+    event VARCHAR,
+    timestamp DateTime64(6, 'UTC'),
+    properties_json VARCHAR,
+)
+ENGINE = MergeTree()
+ORDER BY (toDate(timestamp), event, uuid)
+PARTITION BY toYYYYMM(timestamp)
+```
+
+Each event has an ID,  event type, timestamp, and a JSON representation of event properties. The properties can include the current URL and any other user-defined properties that describe the event (e.g. NPS survey results, person properties, timing data, etc).
+
+This table can be used to store a lot of analytics data and is similar to what we use at PostHog.
+
+If we wanted to query login page pageviews in August, the query would look like this:
+
+```sql
+SELECT count(*)
+FROM events
+WHERE event = '$pageview'
+  AND JSONExtractString(properties_json, '$current_url') = 'https://app.posthog.com/login'
+  AND timestamp >= '2021-08-01'
+  AND timestamp < '2021-09-01'
+```
+
+On a large test dataset this query takes a while complete, while without the url filter the query is almost instant. Adding even more filters just makes the query slower and slower. Let’s dig in why!
+
+## Looking at flamegraphs
+
+ClickHouse has great tools for introspecting queries. Looking at `system.query_log`  we can see that the query:
+
+- Took 3433ms
+- Read 79.17 GiB from disk
+
+To dig even deeper, we can use clickhouse-flamegraph to peek into what the CPU is doing during query execution.
+
+[![Flamegraph](../images/blog/clickhouse-materialized-columns/query-json-extract-CPU.svg)](../images/blog/clickhouse-materialized-columns/query-json-extract-CPU.svg)
+
+From this we can see that the clickhouse node CPU is spending most of its time parsing JSON.
+
+The typical solution would be to extract $current_url to a separate column. This would get rid of the JSON parsing and reduce the amount of data read from disk.
+
+However, in this particular case it wouldn’t work because:
+
+1. The data is passed from users - meaning we’d end up with millions (!) of unique columns
+2. This would complicate live data ingestion a lot, introducing new and exciting race conditions
+
+
+## Enter materialized columns
+
+A less known feature of ClickHouse is the concept of materialized columns. You can add one as follows:
```

Comment:
I'd actually explain what materialized columns are somewhere here, and remove all explanations from the intro. The problem / how it benefits the reader is more important for the intro

Reply:
Done! Let me know if this is better!